In [5]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os 

df = pd.read_csv("../data/cleaned_data.csv") # upload dataset 

In [6]:
# convert time columns to datetime objects
df['start_time'] = pd.to_datetime(df['start_time'])
df['end_time'] = pd.to_datetime(df['end_time'])

# extract useful time features
df['month'] = df['start_time'].dt.month_name()
df['hour'] = df['start_time'].dt.hour

# set month order 
month_order = ['January', 'February', 'March', 'April', 'May', 'June', 
               'July', 'August', 'September', 'October', 'November', 'December']

In [7]:
fig, axes = plt.subplots(4, 3, figsize=(15, 18))
axes = axes.flatten()

for i, month in enumerate(month_order):
    month_data = df[df['month'] == month]
    
    top_states = month_data['state'].value_counts().head(10) # get the top 10 states for that month
    
    sns.barplot(x=top_states.values, y=top_states.index, ax=axes[i])
    
    axes[i].set_xlabel('Number of Counties Affected by Outages')
    axes[i].set_ylabel('')

fig.suptitle('Top 10 States Affected by Outages by Month', fontsize=20, fontweight='bold')

plt.tight_layout()
plt.savefig("../outputs/EDA/top_ten_states_by_month.jpg", format='jpg', dpi=100) 
plt.close()

In [8]:
# plot of frequency of outages by month 
plt.figure(figsize=(6, 4))
sns.countplot(data=df, x='month', order=month_order)

plt.title('Frequency of Outages by Month')
plt.xlabel('Month')
plt.ylabel('Number of Outages')

plt.xticks(rotation=45)

plt.savefig("../outputs/EDA/frequency_of_outages_by_month.jpg", format='jpg', dpi=100)
plt.close()

In [9]:
# sort and count the frequency of each event type
event_counts = df['Event Type'].value_counts()

fig, ax = plt.subplots(figsize=(10, 6))
sns.barplot(x=event_counts.values, y=event_counts.index, ax=ax)

ax.bar_label(ax.containers[0], label_type='edge', padding=5, fontsize=12, weight='bold')

ax.set_title('Distribution of Outages by Event Type', fontsize=18)
ax.set_xlabel('Total Number of Records', fontsize=14)
ax.set_ylabel('Event Type', fontsize=14)

plt.tight_layout()
plt.savefig("../outputs/EDA/distribution_of_outages_by_type.jpg", format='jpg', dpi=100)
plt.close()

In [10]:
# plot the distribution of the time of day outage began
hour_counts = df["hour"].value_counts().sort_index()
hour_counts.plot(kind="bar")

plt.xlabel("Hour of Day")
plt.ylabel("Number of Outages")
plt.title("Outage Count by Hour")
plt.savefig("../outputs/EDA/distribution_of_outages_by_hour.jpg", format='jpg', dpi=100) 
plt.close()

In [12]:
# plot the distribution of the duration of outages 
plt.figure(figsize=(8, 6))
sns.boxplot(x=df['duration'], color='skyblue')

plt.title('Boxplot of Duration', fontsize=16)
plt.xlabel('Duration', fontsize=14)

plt.tight_layout()
plt.savefig("../outputs/EDA/boxplot_duration.jpg", format='jpg', dpi=100)
plt.close()

In [14]:
df['duration'].describe()

count    77341.000000
mean         6.339713
std         15.464254
min          0.250000
25%          0.500000
50%          1.500000
75%          3.750000
max        193.250000
Name: duration, dtype: float64

This tells us that the mean duration of the outages is ~6 hours, with a large range in duration however, from 15 minutes to ~193 hours. 